### PR TITLE
Created a method of lighting control

### PR DESCRIPTION
**Pull request by ghost**
Added TwoWayComm as a method of communicating with serial devices
Added unit tests
Added DisplayLighting as a continuous task to change LED color

https://app.asana.com/0/219272915179160/250106024468355

### CODE REVIEW

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/DisplayLighting.scala [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  override def onStart(): Unit = {
+    comm.pushDataToQueue(rgbToData(r, g, b))
+  }
+
+  override def onEnd(): Unit = {
+    comm.pushDataToQueue(rgbToData(0, 0, 0))
+  }
+
+  private def rgbToData(r: Int, g: Int, b: Int): String = {
+    val rR = {
```
Comment (Vincent-Belliard) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/DisplayLighting.scala]: you can replace all this function by:
"%03d%03d%03d".format(r, g, b)

Comment (shadaj) [reply lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/DisplayLighting.scala]: Or we can use string interpolation! You can use `f"$r%03d$g%03d$b%03d"`

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/DisplayLighting.scala [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package com.lynbrookrobotics.potassium.lighting
+
+import com.lynbrookrobotics.potassium.tasks.ContinuousTask
+
+/**
+  * Created by Timothy on 1/13/2017.
+  * Continuous Task that you feed a communicator and an rgb value.
+  * Defaults to black onEnd
+  */
+class DisplayLighting(r: Int, g: Int, b: Int, comm: TwoWayComm) extends ContinuousTask {
```
Comment (PhilipAxelrod) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/DisplayLighting.scala]: Probably want to make this a Color? Also, should probably be a case class

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/DisplayLighting.scala [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  require(r > -1 && r < 256 && g > -1 && g < 256 && b > -1 && b < 256, "Unacceptable color value")
+
+  override def onStart(): Unit = {
+    comm.pushDataToQueue(f"$r%03d$g%03d$b%03d")
```
Comment (PhilipAxelrod) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/DisplayLighting.scala]: Can you clarify/ add comments? Because this is kind of scary

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/DisplayLighting.scala [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  }
+
+  override def onEnd(): Unit = {
+    comm.pushDataToQueue("000000000")
```
Comment (PhilipAxelrod) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/DisplayLighting.scala]: See above

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+import gnu.io.{CommPortIdentifier, SerialPort}
+
+import scala.collection.mutable
+
+/**
+  * Created by Timothy on 1/14/2017.
+  * Communication protocol for communicating with serial devices.
+  */
+class TwoWayComm{
```
Comment (PhilipAxelrod) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit: space before {

Comment (PhilipAxelrod) [reply lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Also what protocol is this?

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+  /**
+    * DO NOT USE
+    */
+  val dataQueue = mutable.Queue[Int]()
+  /**
+    * DO NOT USE
+    */
+  val logQueue = mutable.Queue[String]()
```
Comment (PhilipAxelrod) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Why is this here? Also, add deprecated annotation

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+      commPort match {
+        case serialPort: SerialPort =>
+          serialPort.setSerialPortParams(115200, SerialPort.DATABITS_8, SerialPort.STOPBITS_1, SerialPort.PARITY_NONE)
+
+          val in = serialPort.getInputStream
+          val out = serialPort.getOutputStream
+
+          new Thread(new SerialReader(in)).start()
+          new Thread(new SerialWriter(out)).start()
+        case default => logQueue.enqueue("something went wrong...")
```
Comment (PhilipAxelrod) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Plss add something more descriptive

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+    }
+  }
+
+  class SerialReader(val in: InputStream) extends Runnable {
+    override def run(): Unit = {
+      val bucket = new Array[Byte](1024)
+      try {
+        var len = -1;
+        while ({
+          len = this.in.read(bucket)
```
Comment (PhilipAxelrod) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: this.in.read(bucket) > -1

Comment (PhilipAxelrod) [reply lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Clarify what len is

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+      } catch {
+        case e: Exception => e.printStackTrace()
+      }
+    }
+  }
+
+  class SerialWriter(val out: OutputStream) extends Runnable {
+    override def run(): Unit = {
+      try {
+        var c = 0
```
Comment (PhilipAxelrod) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Please clarify

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+        while ( {
+          if(dataQueue.nonEmpty) {
+            c = dataQueue.dequeue()
```
Comment (PhilipAxelrod) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit: No parenthesis

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+          }else{
```
Comment (PhilipAxelrod) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit: space

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+  }
+
+  /**
+    * Pushes a number to the data queue to be processed
+    *
+    * @param data integer to be sent
+    */
+  def pushDataToQueue(data: String): Unit = {
+    for(c <- data){
+      dataQueue.enqueue(c.toInt)
```
Comment (PhilipAxelrod) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Try dataQueue.foreach

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+    }
+  }
+
+  /**
+    * Pulls data from the queue with data created by the serial device
+    *
+    * @return First point of data on the Queue
+    */
+  def pullLog(): String = {
```
Comment (PhilipAxelrod) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit: Parenthesis

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+
+  /**
+    * Finds out the port of your system
+    *
+    * @return the port to use in connect
+    */
+  def systemPort(): String = {
+    val portEnum = CommPortIdentifier.getPortIdentifiers
+    while(portEnum.hasMoreElements){
+      val commPortId = portEnum.nextElement().asInstanceOf[CommPortIdentifier]
```
Comment (PhilipAxelrod) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: nit: parenthesis

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+        }
+    }
+    throw new RuntimeException
+  }
+
+  /**
+    * For testing purposes only
+    * @return testing SerialReader
+    */
+  def serialReader(): SerialReader = {
```
Comment (PhilipAxelrod) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit Parenthesis

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+    * Finds out the port of your system
+    *
+    * @return the port to use in connect
+    */
+  def systemPort(): String = {
+    val portEnum = CommPortIdentifier.getPortIdentifiers
+    while(portEnum.hasMoreElements){
+      val commPortId = portEnum.nextElement().asInstanceOf[CommPortIdentifier]
+      for(n <- PORT_NAMES)
+        if(commPortId.getName == n){
```
Comment (PhilipAxelrod) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: PORT_NAMES.filter( _ == n)
if PORT_NAMES is a list

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+
+  /**
+    * Pulls data from the queue with data created by the serial device
+    *
+    * @return First point of data on the Queue
+    */
+  def pullLog(): String = {
+    if (logQueue.nonEmpty) {
+      logQueue.dequeue()
+    }else{
```
Comment (PhilipAxelrod) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit space

**Review by PhilipAxelrod — DISMISSED** [review]

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/DisplayLighting.scala [hunk 18]
```diff
@@ -0,0 +1,20 @@
+package com.lynbrookrobotics.potassium.lighting
+
+import com.lynbrookrobotics.potassium.tasks.ContinuousTask
+
+/**
+  * Created by Timothy on 1/13/2017.
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/DisplayLighting.scala]: Can you remove the generated "Created by" comments?

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+  * Created by Timothy on 1/14/2017.
+  * RXTX communication protocol for communicating with serial devices.
+  */
+class TwoWayComm {
+
+  /**
+    * DO NOT USE
+    */
+    @deprecated
+  val dataQueue = mutable.Queue[Int]()
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Can you just make this `private[lighting]` so that you don't need the comment?

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+      }
+    }
+  }
+
+  class SerialWriter(val out: OutputStream) extends Runnable {
+    override def run(): Unit = {
+      try {
+        var dataToSend = 0
+        while ( {
+          if(dataQueue.nonEmpty) {
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Can you put this as a method inside this method (that's possible in Scala!).

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala (hunk 12 above)
Comment (shadaj) [reply lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: 👍

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+
+import gnu.io.{CommPortIdentifier, SerialPort}
+
+import scala.collection.mutable
+
+/**
+  * RXTX communication protocol for communicating with serial devices.
+  */
+class TwoWayComm {
+
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit: extra lines

File: lighting/src/test/scala/com/lynbrookrobotics/potassium/lighting/LightingTest.scala [hunk 22]
```diff
@@ -0,0 +1,36 @@
+package com.lynbrookrobotics.potassium.lighting
+
+import java.io.{InputStream, OutputStream}
+
+import com.lynbrookrobotics.potassium.tasks.Task
+import org.scalatest.FunSuite
+
+//TODO: Implement Mokito
+class LightingTest extends FunSuite{
```
Comment (shadaj) [inline lighting/src/test/scala/com/lynbrookrobotics/potassium/lighting/LightingTest.scala]: Nit: space before {

File: lighting/src/test/scala/com/lynbrookrobotics/potassium/lighting/LightingTest.scala [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+  val dummy = new TwoWayComm(){
```
Comment (shadaj) [inline lighting/src/test/scala/com/lynbrookrobotics/potassium/lighting/LightingTest.scala]: Nit: space before {

File: lighting/src/test/scala/com/lynbrookrobotics/potassium/lighting/LightingTest.scala [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+    override def pushDataToQueue(data: String){
```
Comment (shadaj) [inline lighting/src/test/scala/com/lynbrookrobotics/potassium/lighting/LightingTest.scala]: Nit: space before {

File: lighting/src/test/scala/com/lynbrookrobotics/potassium/lighting/LightingTest.scala [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+      logQueue.enqueue(data)
+    }
+    def logQueueEmpty():Boolean = logQueue.isEmpty
```
Comment (shadaj) [inline lighting/src/test/scala/com/lynbrookrobotics/potassium/lighting/LightingTest.scala]: Nit: space before `Boolean`

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 26]
```diff
@@ -0,0 +1,154 @@
...
+    }
+  }
+
+  class SerialWriter(val out: OutputStream) extends Runnable {
+    override def run(): Unit = {
+      var dataToSend = 0
+      def checkData:Boolean = {
+        if(dataQueue.nonEmpty) {
+          dataToSend = dataQueue.dequeue
+        }else{
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit: space after }

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 27]
```diff
@@ -0,0 +1,154 @@
...
+          dataToSend = 0
+        }
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit: line after }

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 28]
```diff
@@ -0,0 +1,154 @@
...
+        dataToSend > -1
+      }
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit: line after }

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 29]
```diff
@@ -0,0 +1,154 @@
...
+        case e: Exception => e.printStackTrace()
+      }
+    }
+  }
+
+  class SerialWriter(val out: OutputStream) extends Runnable {
+    override def run(): Unit = {
+      var dataToSend = 0
+      def checkData:Boolean = {
+        if(dataQueue.nonEmpty) {
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit: space before (

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 30]
```diff
@@ -0,0 +1,154 @@
...
+    override def run(): Unit = {
+      val bucket = new Array[Byte](1024)
+      try {
+        var len = -1; //Number of bytes read
+        def checkLength: Boolean ={
+          {
+            len = this.in.read(bucket)
+            len > -1
+          }
+        }
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit: line after }

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 31]
```diff
@@ -0,0 +1,154 @@
...
+      }
+    }
+  }
+
+  class SerialReader(val in: InputStream) extends Runnable {
+    override def run(): Unit = {
+      val bucket = new Array[Byte](1024)
+      try {
+        var len = -1; //Number of bytes read
+        def checkLength: Boolean ={
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit: space before {

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 32]
```diff
@@ -0,0 +1,154 @@
...
+class TwoWayComm {
+
+
+  protected val dataQueue = mutable.Queue[Int]()
+
+  protected val logQueue = mutable.Queue[String]()
+
+  /**
+    * Connect to serial device
+    *
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit: extra line

**Review by shadaj — CHANGES_REQUESTED** [review]
Just some nits. Also, where is the Arduino code for receiving data? It should also be checked in under src/main/arduino

**Comment by shadaj** [comment]
Looks like you have conflicts with `master`. Can you rebase?

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/LightingComponent.scala [hunk 33]
```diff
@@ -0,0 +1,62 @@
...
+    */
+  override def applySignal(signal: List[Color]): Unit = {
+    def helper(color: Color): String = {
+      val r = color.getRed
+      val g = color.getGreen
+      val b = color.getBlue
+      f"$r%03d$g%03d$b%03d"
+    }
+    var acc = ""
+    signal.foreach(acc += helper(_))
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/LightingComponent.scala]: Can you replace with `signal.map(helper).mkString`?

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/LightingComponent.scala [hunk 34]
```diff
@@ -0,0 +1,62 @@
...
+        *
+        * @return the latest value of the signal
+        */
+      override def get: List[Color] = {
+        if (colorStore == color && up) {
+          up = false
+        } else if (colorStore == Color.BLACK && !up) {
+          up = true
+        }
+        if(up) {
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/LightingComponent.scala]: Space before `(`

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/LightingComponent.scala [hunk 35]
```diff
@@ -0,0 +1,62 @@
...
+          colorStore = colorStore.brighter()
+          List.fill(numLEDs)(colorStore)
+        }else {
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/LightingComponent.scala]: Space after `}`

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/LightingComponent.scala [hunk 36]
```diff
@@ -0,0 +1,62 @@
...
+          colorStore = colorStore.darker()
+          List.fill(numLEDs)(colorStore)
+        }
+      }
+    }
+    signal
+  }
+
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/LightingComponent.scala]: Extra line

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 37]
```diff
@@ -0,0 +1,132 @@
...
+  protected lazy val portIdentifier = CommPortIdentifier.getPortIdentifier(portName)
+
+  /**
+    * Connect to serial device
+    */
+  def connect(): Unit = {
+    if (portIdentifier.isCurrentlyOwned) {
+      logQueue.enqueue("Error: Port currently in use")
+    }
+    else {
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Else should be on previous line.

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 38]
```diff
@@ -0,0 +1,132 @@
...
+      val commPort = portIdentifier.open(this.getClass().getName(), 2000)
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Just use `getClass()` instead of `this.`

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 39]
```diff
@@ -0,0 +1,132 @@
...
+    def run {
+      val buffer: Array[Byte] = new Array[Byte](1024)
+      var len: Int = -1
+      try {
+        def helper(): Boolean = {
+          len = in.read(buffer)
+          len > -1
+        }
+        while (helper()) {
+          {
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: No need for these wrapping braces

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 40]
```diff
@@ -0,0 +1,132 @@
...
+            logQueue.enqueue(new String(buffer, 0, len))
+          }
+        }
+      }
+      catch {
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: `catch` should be on previous line

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 41]
```diff
@@ -0,0 +1,132 @@
...
+
+  class SerialWriter(var out: OutputStream) extends Runnable {
+    def run {
+        var c: Int = 0
+        def helper():Boolean = {
+          c = dataQueue.dequeue()
+          c > -1
+        }
+        while (helper()) {
+          {
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: No need for wrapping braces

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 42]
```diff
@@ -0,0 +1,132 @@
...
+    }
+  }
+
+  class SerialWriter(var out: OutputStream) extends Runnable {
+    def run {
+        var c: Int = 0
+        def helper():Boolean = {
+          c = dataQueue.dequeue()
+          c > -1
+        }
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala:77]: Line after the method

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 43]
```diff
@@ -0,0 +1,132 @@
...
+
+  class SerialReader(var in: InputStream) extends Runnable {
+    def run {
+      val buffer: Array[Byte] = new Array[Byte](1024)
+      var len: Int = -1
+      try {
+        def helper(): Boolean = {
+          len = in.read(buffer)
+          len > -1
+        }
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala:58]: Line after the method

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 44]
```diff
@@ -0,0 +1,132 @@
...
+    "COM3");
+
+  /**
+    * Finds out the port of your system
+    *
+    * @return the port to use in connect
+    */
+  def systemPort: String = {
+    val portEnum = CommPortIdentifier.getPortIdentifiers
+    while(portEnum.hasMoreElements){
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Space before `(`

File: lighting/src/test/scala/com/lynbrookrobotics/potassium/lighting/LightingTest.scala [hunk 45]
```diff
@@ -0,0 +1,70 @@
...
+    dummyFail.connect()
+    assertResult("Error: Port currently in use")(dummyFail.pullLog)
+  }
+
+  test("Test Pattern Matching"){
+    val dummyFail = new DummyFail(false)
+    dummyFail.connect()
+    assertResult("Error: Comm port not a serial port")(dummyFail.pullLog)
+  }
+  */
```
Comment (shadaj) [inline lighting/src/test/scala/com/lynbrookrobotics/potassium/lighting/LightingTest.scala]: Extra `*/`

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/DisplayLighting.scala [hunk 46]
```diff
@@ -0,0 +1,21 @@
...
+
+import com.lynbrookrobotics.potassium.Signal
+import com.lynbrookrobotics.potassium.tasks.ContinuousTask
+
+/**
+  * Continuous Task that you feed a communicator and an rgb value.
+  * Defaults to black onEnd
+  */
+class DisplayLighting(signal: Signal[List[Color]], lightingComponent: LightingComponent) extends ContinuousTask {
+
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/DisplayLighting.scala]: Nit: extra line

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/LightingComponent.scala [hunk 47]
```diff
@@ -0,0 +1,39 @@
...
+import java.awt.Color
+
+import com.lynbrookrobotics.potassium.clock.Clock
+import com.lynbrookrobotics.potassium.{Component, PeriodicSignal, Signal}
+import squants.time.Milliseconds
+
+import scala.util.Failure
+
+class LightingComponent(numLEDs: Int, comm: TwoWayComm)(implicit clock: Clock) extends Component[List[Color]](Milliseconds(20)) {
+
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/LightingComponent.scala]: Nit: extra line

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/LightingComponent.scala [hunk 48]
```diff
@@ -0,0 +1,39 @@
...
+    * @param signal the signal value to act on
+    */
+  override def applySignal(signal: List[Color]): Unit = {
+    def helper(i: Int, color: Color): String = {
+      val r = color.getRed
+      val g = color.getGreen
+      val b = color.getBlue
+      f"$i%03d$r%03d$g%03d$b%03d"
+    }
+    if (comm.isConnected) {
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/LightingComponent.scala]: Nit: should have line above

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 49]
```diff
@@ -0,0 +1,133 @@
...
+        case serialPort: SerialPort =>
+          serialPort.setSerialPortParams(115200, SerialPort.DATABITS_8, SerialPort.STOPBITS_1, SerialPort.PARITY_NONE)
+          val in = serialPort.getInputStream
+          val out = serialPort.getOutputStream
+          startThreads(in, out)
+          isConnected = true
+        case _ => logQueue.enqueue("Error: Comm port not a serial port")
+      }
+    }
+    isConnected
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit: should have line above

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 50]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  protected def startThreads(in: InputStream, out: OutputStream): Unit ={
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit: space before {

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 51]
```diff
@@ -0,0 +1,133 @@
...
+          c > -1
+        }
+
+        while (helper()) {
+          this.out.write(c)
+        }
+    }
+  }
+
+
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit: extra line

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 52]
```diff
@@ -0,0 +1,133 @@
...
+  /**
+    * Finds out the port of your system
+    *
+    * @return the port to use in connect
+    */
+  def systemPort: String = {
+    val portEnum = CommPortIdentifier.getPortIdentifiers
+    while (portEnum.hasMoreElements) {
+      val commPortId = portEnum.nextElement.asInstanceOf[CommPortIdentifier]
+      PORT_NAMES.filter(_ == commPortId.getName)
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: This doesn't do anything, are you trying to find the identifier that is in your list? In that case you can just do `portEnum.(there's something to get a list but I don't remember).find(id => PORT_NAMES.contains(id.getName)).getOrElse("COM3")`

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 53]
```diff
@@ -0,0 +1,133 @@
...
+    */
+  def pullLog: String = {
+    if (logQueue.nonEmpty) {
+      logQueue.dequeue
+    }else {
+      "No data to show"
+    }
+  }
+
+  private val PORT_NAMES = List[String](
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Scala style for constants is capitalcase, so should be `PortNames`

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala [hunk 54]
```diff
@@ -0,0 +1,133 @@
...
+    while (portEnum.hasMoreElements) {
+      val commPortId = portEnum.nextElement.asInstanceOf[CommPortIdentifier]
+      PORT_NAMES.filter(_ == commPortId.getName)
+    }
+    "COM3"
+  }
+
+  @tailrec
+  final def dumpLog: Unit = {
+    if(logQueue.nonEmpty) {
```
Comment (shadaj) [inline lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: Nit: space before (

File: lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala (hunk 52 above)
Comment (ghost) [reply lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: This should work with java enumerations

Comment (ghost) [reply lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/TwoWayComm.scala]: I think you are mistaking the enumeration for a scala one

**Review by shadaj — APPROVED** [review]
LGTM! Thanks for persisting through the (quite long) review process!
